rir/src/compiler/opt/assumptions.cpp

```diff
@@ -0,0 +1,58 @@
+#include "../pir/pir_impl.h"
+#include "../translations/pir_translator.h"
+#include "../util/cfg.h"
+#include "../util/visitor.h"
+
+#include "R/r.h"
+#include "pass_definitions.h"
+
+#include <unordered_map>
+#include <unordered_set>
+
+namespace rir {
+namespace pir {
+
+void OptimizeAssumptions::apply(RirCompiler&, Closure* function,
+                                LogStream&) const {
+    CFG cfg(function);
+    Visitor::run(function->entry, [&](BB* bb) {
+        auto ip = bb->begin();
+        while (ip != bb->end()) {
+            auto next = ip + 1;
+            auto instr = *ip;
+
+            if (auto assume = Assume::Cast(instr)) {
+                auto cp = assume->checkpoint();
+                auto bb = cp->bb();
+                // We are trying to group multiple assumes into the same
```

Comment:
I was a little confused by the comment. `moving up linear sequence of bbs` made me think we are moving bbs.
I would say: `We are trying to group multiple assumes into the same checkpoint by finding for each assume the topmost compatible checkpoint.`
    
Then, again, just brainstorming...I have the impression this is becoming a kind of pattern we are facing many times. 
Maybe we can add to cfg a function `findTopmostComplyingWith(Instruction inst, lambda condition)` and iterate it till we reach a general version. I mean the code here would be much easier to read if we replace the whole while with